### PR TITLE
Assignment6 lambda expressions: With Jessica Ho

### PR DESCRIPTION
**Pull request by jacobnewcomb**
With @JessicaH46

### CODE REVIEW

File: Calculate/Calculate/Program.cs [hunk 1]
```diff
@@ -0,0 +1,62 @@
+﻿using System;
+namespace Calculate
+{
+	public class Program
+	{
+		public Action<string> WriteLine { get; init; } = Console.WriteLine;
+		public Func<string> ReadLine { get; init; } = Console.ReadLine;
+
+		public Program(Action<string> writeLine, Func<string> readLine)
```
Comment (GeorgeKoralski007) [inline Calculate/Calculate/Program.cs:9]: No need to make a constructor. 
Set the default behavior for the WriteLine and ReadLine properties to invoke System.Console versions of the methods and add an empty default constructor.

**Comment by GeorgeKoralski007** [comment]
I will review

File: Calculate/Calculate/Calculate.csproj [hunk 2]
```diff
@@ -0,0 +1,12 @@
+<Project Sdk="Microsoft.NET.Sdk">
+
+  <PropertyGroup>
+    <OutputType>Exe</OutputType>
+    <TargetFramework>net7.0</TargetFramework>
+    <ImplicitUsings>enable</ImplicitUsings>
+    <Nullable>enable</Nullable>
+	<EnableNETAnalyzers>true</EnableNETAnalyzers>
+	<LangVersion>11.0</LangVersion>
+  </PropertyGroup>
```
Comment (GeorgeKoralski007) [inline Calculate/Calculate/Calculate.csproj:10]: Place all shared project properties into a Directory.Build.Props file.

File: Calculate/Calculate/Calculator.cs [hunk 3]
```diff
@@ -0,0 +1,33 @@
+﻿namespace Calculate
+{
+    public class Calculator
+    {
+        public static Func<int, int, int> Add = (a, b) => a + b;
+        public static Func<int, int , int> Subtract = (a, b) => a - b;
+        public static Func<int, int, int> Multiply = (a, b) => a * b;
+        public static Func<int, int, int> Divide = (a, b) => { if (b != 0) return a / b; else throw new DivideByZeroException(); };
```
Comment (GeorgeKoralski007) [inline Calculate/Calculate/Calculator.cs:8]: No need to throw DivideByZeroExeception because the division operator already does that. Test will show you that it already throws an exception.

File: Calculate/Calculate/Program.cs [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+			ReadLine = readLine;
+		}
+
+		public Program() { } //empty constr
+
+		static void Main(string[] args)
+		{
+			var program = new Program();
+			{
+				program.WriteLine("WriteLine test");
```
Comment (GeorgeKoralski007) [inline Calculate/Calculate/Program.cs:21]: This code should be in the Test, not in the main program.

File: Calculate/Calculate/Program.cs [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+
+			while(exit == false)
+			{
+				Console.WriteLine("Please input function: (format) x [opertor] y, or 'q' to quit program");
+
+				string userInput = Console.ReadLine().Trim();
+
+				if (string.IsNullOrWhiteSpace(userInput) || string.IsNullOrEmpty(userInput))
+				{
+					Console.WriteLine("Invalid Input, please try again");
```
Comment (GeorgeKoralski007) [inline Calculate/Calculate/Program.cs:40]: When you detect invalid input, you need to call "continue;" to reset the loop

File: Calculate/Calculate/Program.cs [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+				}
+
+				if (Calculator.TryCalculate(userInput, out int x))
+				{
+					program.WriteLine(x.ToString());
+				}
+
+				else if (userInput.Equals("q"))
```
Comment (GeorgeKoralski007) [inline Calculate/Calculate/Program.cs:48]: Its better to check for "q" first so you don't have to call TryCalculate

File: Calculate/Calculate/Program.cs [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+				{
+					exit = true;
```
Comment (GeorgeKoralski007) [inline Calculate/Calculate/Program.cs:50]: Instead of using an exit flag, you can call "break;" to exit the loop

File: Calculate/Calculate/Program.cs [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+			}
+
+			Calculator calculator = new();
+			//calculator.TryCalculate(program.WriteLine, program.ReadLine);
+
+			bool exit = false;
+
+			while(exit == false)
+			{
+				Console.WriteLine("Please input function: (format) x [opertor] y, or 'q' to quit program");
```
Comment (GeorgeKoralski007) [inline Calculate/Calculate/Program.cs:34]: You need to use program.WriteLine in all placed like you did on line 24.

**Comment by j0shh3ss** [comment]
Reviewing:
Create a Console project called "Calculate.". ✔
Define a Program Class✔
Define two init-only setter properties, WriteLine and ReadLine, that contain delegates for writing a line of text and reading a line of text respectively ✔
Write a test that sets these properties at construction time and then invokes the properties and verifies the expected behavior occurs. ❌
Set the default behavior for the WriteLine and ReadLine properties to invoke System.Console versions of the methods and add an empty default constructor. ✔
Define a Calculator class ✔
Define static Add, Subtract, Multiple, and Divide methods that have two parameters and return a third parameter. ✔
Define a read-only property, MathematicalOperations, of type System.Collections.Generics.IReadOnlyDictionary<TKey,TValue> that:
is initialized to a System.Collections.Generics.Dictionary<<TKey,TValue> instance that. ✔
Uses char for the key corresponding to the operators +, -, *, and /. ✔
Has values that correspond with the Add, Subtract, Multiple, and Divide methods. ✔
Implement a TryCalculate method following "TryParse" pattern ✔
Valid calculation expressions include such strings as "3 + 4", "42 - 2", etc. ✔
If there is no whitespace around the operator, you can assume the calculation is invalid and return false. Similarly if the operands are not integers. ✔
Use string.Split(), pattern matching, logical and operators to parse the string in their entirety ✔
Index into the MathematicalOperations method using the operator parsed during pattern matching to find the corresponding implementation and invoke it. ✔
Implement the Program class to instantiate the calculator and invoke it based on user input from the console. ✔
Be sure to use the WriteLine/ReadLine properties on Program for testing the input and output of your program. ✔

Remember to add test files testing all properites in Calculator and Program. Also be sure to do what George said above. Looks like you're in a good spot!

**Comment by GeorgeKoralski007** [comment]
Create a Console project called "Calculate.". ✔
Define a Program Class
Define two init-only setter properties, WriteLine and ReadLine, that contain delegates for writing a line of text and reading a line of text respectively ✔
Write a test that sets these properties at construction time and then invokes the properties and verifies the expected behavior occurs. ❌
Set the default behavior for the WriteLine and ReadLine properties to invoke System.Console versions of the methods and add an empty default constructor. ✔
Define a Calculator class ✔
Define static Add, Subtract, Multiple, and Divide methods that have two parameters and return a third parameter. ✔
Define a read-only property, MathematicalOperations, of type System.Collections.Generics.IReadOnlyDictionary<TKey,TValue> that:
is initialized to a System.Collections.Generics.Dictionary<<TKey,TValue> instance that. ✔
Uses char for the key corresponding to the operators +, -, *, and /. ✔
Has values that correspond with the Add, Subtract, Multiple, and Divide methods. ✔
Implement a TryCalculate method following "TryParse" pattern ✔
Valid calculation expressions include such strings as "3 + 4", "42 - 2", etc. ✔
If there is no whitespace around the operator, you can assume the calculation is invalid and return false. Similarly if the operands are not integers. ✔
Use string.Split(), pattern matching, logical and operators to parse the string in their entirety ✔
Index into the MathematicalOperations method using the operator parsed during pattern matching to find the corresponding implementation and invoke it. ✔
Implement the Program class to instantiate the calculator and invoke it based on user input from the console. ✔
Be sure to use the WriteLine/ReadLine properties on Program for testing the input and output of your program. ❌Replace Console.WriteLine with program.WriteLine
Same goes for ReadLine
Fundamentals
Place all shared project properties into a Directory.Build.Props file. ❌
Place all shared project items into a Directory.Build.targets file.
nullable reference types is enabled ✔
Ensure that you turn on code analysis for all projects(EnableNETAnalyzers) ✔
Set LangVersion and the TargetFramework to the latest released versions available (preview versions optional) ✔
and enabled .NET analyzers for both projects ✔
For this assignment, always use Assert.AreEqual<T>() (the generic version) ❌
All of the above should be unit tested ❌
Choose simplicity over complexity ✔

Need to create unit tests.

**Review by GeorgeKoralski007 — CHANGES_REQUESTED** [review]
Please review

**Review by StrangeRanger — COMMENTED** [review]
## Instructions

- Create a *Console* project called "Calculate.". ✔
- Define a Program Class
  - Define two init-only setter properties, `WriteLine` and `ReadLine`, that contain delegates for writing a line of text and reading a line of text respectively ✔
  - Write a test that sets these properties at construction time and then invokes the properties and verifies the expected behavior occurs. ❌
  - Set the default behavior for the `WriteLine` and `ReadLine` properties to invoke `System.Console` versions of the methods and add an empty default constructor. ✔
- Define a Calculator class ✔
  - Define static `Add`, `Subtract`, `Multiple`, and `Divide` methods with two parameters and return a third parameter. ✔
  - Define a read-only property, `MathematicalOperations`, of type `System.Collections.Generics.IReadOnlyDictionary<TKey,TValue>` that:
    - is initialized to a `System.Collections.Generics.Dictionary<<TKey,TValue>` instance that. ✔
      - Uses `char` for the key corresponding to the operators +, -, *, and /. ✔
      - Has values that correspond with the Add, Subtract, Multiple, and Divide methods. ✔
  - Implement a `TryCalculate` method following the "TryParse" pattern ✔ ???
    - Valid `calculation` expressions include such strings as "3 + 4", "42 - 2", etc. ✔
    - If there is no whitespace around the operator, you can assume the `calculation` is invalid and return false. Similarly, if the operands are not integers. ✔
    - Use `string.Split()`, pattern matching, logical, and operators to parse the string in its entirety ✔
       - ***REVIEW NOTES: Ideally we want to use some form of pattern matching similar to what we were shown/introduced to, I believe, last Thursday.***
    - Index into the `MathematicalOperations` method using the operator parsed during pattern matching to find the corresponding implementation and invoke it. ✔
- Implement the Program class to instantiate the calculator and invoke it based on user input from the console. ✔
- Be sure to use the `WriteLine`/`ReadLine` properties on `Program` for testing the input and output of your program. ❌
  - ***REVIEW NOTES: While you did use `WriteLine` and `ReadLine` where appropriate, you should replace all instances of `Console.WriteLine` with your implemented version of `WriteLine`.

## Fundamentals

- Place all shared project properties into a `Directory.Build.Props` file.
- Place all shared project items into a `Directory.Build.targets` file.
- nullable reference types is enabled  ✔
- Ensure that you turn on code analysis for all projects(EnableNETAnalyzers)  ✔
- Set LangVersion and the TargetFramework to the latest released versions available (preview versions optional)   ✔
- and enabled .NET analyzers for both projects ❌✔
  - ***REVIEW NOTES: While technically a check mark, a test project should exist where you'd additionally enable the .NET analyzers.***
- For this assignment, always use `Assert.AreEqual<T>()` (the generic version)  ❌
- All of the above should be unit tested ❌
- Choose simplicity over complexity ✔

## Notes

For the most part, the program seems to do what it needs to be done, but it feels like you got to a certain point in your coding and just said, "screw it, it's good enough" and stopped entirely. Maybe that's because there was never supposed to be a whole lot of code in the first place, but it just feels incomplete for some reason. So realistically, you can probably just ignore this comment. I just wanted to add it as a just in case...

With that said, the only other thing is that you want to make sure you provide unit testing. At this point, it should be well known that we write our tests before writing our actual code if at least not in parallel with each other. If you did so, you'd have realized that you didn't need all those if else statements in the try calculate to ensure that a valid equation was provided.

**Comment by JessicaH46** [comment]
> ## Notes
> For the most part, the program seems to do what it needs to be done, but it feels like you got to a certain point in your coding and just said, "screw it, it's good enough" and stopped entirely. Maybe that's because there was never supposed to be a whole lot of code in the first place, but it just feels incomplete for some reason. So realistically, you can probably just ignore this comment. I just wanted to add it as a just in case...
> 
> With that said, the only other thing is that you want to make sure you provide unit testing. At this point, it should be well known that we write our tests before writing our actual code if at least not in parallel with each other. If you did so, you'd have realized that you didn't need all those if else statements in the try calculate to ensure that a valid equation was provided.

Thank you for reviewing our code, like you mentioned our code was still in progress and we were just trying to get what we had at the time in for our initial submission. I also had some issues in my ProgramTests that I couldn't fix before the deadline so I just left some things that we needed for the key components in my Program as more of a placeholder. Mainly just had a lot of things on my plate and I was going to go back and fix it, not really saying "screw it". I see how it did look incomplete, because it was and thank you for the reminder on the WriteLine, I completely forgot to change that! Will make sure to go back and edit them.

File: Calculate/Calculate/Program.cs (hunk 8 above)
Comment (JessicaH46) [reply Calculate/Calculate/Program.cs]: Thank you for the reminder!

File: Calculate/Calculate/Program.cs (hunk 7 above)
Comment (JessicaH46) [reply Calculate/Calculate/Program.cs]: Thank you, changed it! That was a lot simpler, thank you for the suggestion!

File: Calculate/Calculate/Program.cs (hunk 4 above)
Comment (JessicaH46) [reply Calculate/Calculate/Program.cs:21]: I was having some problems with the tests originally and left it here as a placeholder, thank you for the reminder!

File: Calculate/Calculate/Program.cs [hunk 9]
```diff
@@ -0,0 +1,62 @@
...
+					continue;
+				}
+
+                if (userInput.Equals("q"))
+                {
+					break;
+                }
+
+                else if (Calculator.TryCalculate(userInput, out int x))
+				{
```
Comment (lsobtafo) [inline Calculate/Calculate/Program.cs:48]: The TryCalculate method returns a boolean indicating whether the calculation was successful or not, and stores the result in an out parameter. However, it think it would be more consistent with other methods in the class to return the result directly, and throw an exception if the calculation fails. This would make the code more consistent and easier to understand. You might also want to consider using a switch statement instead of a dictionary to map operators to functions, which would make the code more readable and easier to maintain.

**Comment by lsobtafo** [comment]
Reviewing:
Create a Console project called "Calculate.". ✔
Define a Program Class✔
Define two init-only setter properties, WriteLine and ReadLine, that contain delegates for writing a line of text and reading a line of text respectively ✔
Write a test that sets these properties at construction time and then invokes the properties and verifies the expected behavior occurs. ✔
Set the default behavior for the WriteLine and ReadLine properties to invoke System.Console versions of the methods and add an empty default constructor. ✔
Define a Calculator class ✔
Define static Add, Subtract, Multiple, and Divide methods that have two parameters and return a third parameter. ✔
Define a read-only property, MathematicalOperations, of type System.Collections.Generics.IReadOnlyDictionary<TKey,TValue> that:
is initialized to a System.Collections.Generics.Dictionary<<TKey,TValue> instance that. ✔
Uses char for the key corresponding to the operators +, -, *, and /. ✔
Has values that correspond with the Add, Subtract, Multiple, and Divide methods. ✔
Implement a TryCalculate method following "TryParse" pattern ✔
Valid calculation expressions include such strings as "3 + 4", "42 - 2", etc. ✔
If there is no whitespace around the operator, you can assume the calculation is invalid and return false. Similarly if the operands are not integers. ✔
Use string.Split(), pattern matching, logical and operators to parse the string in their entirety ✔
Index into the MathematicalOperations method using the operator parsed during pattern matching to find the corresponding implementation and invoke it. ✔
Implement the Program class to instantiate the calculator and invoke it based on user input from the console. ✔
Be sure to use the WriteLine/ReadLine properties on Program for testing the input and output of your program. ✔

File: Calculate/Calculate.Tests/ProgramTests.cs [hunk 10]
```diff
@@ -0,0 +1,29 @@
+﻿namespace Calculate.Tests
+{
+    [TestClass]
+    public class ProgramTests
```
Comment (mmwoodfo) [inline Calculate/Calculate.Tests/ProgramTests.cs:4]: These test seem to be failing

File: Calculate/Calculate/Calculator.cs [hunk 11]
```diff
@@ -0,0 +1,33 @@
...
+            { '+', Add },
+            { '-', Subtract },
+            { '*', Multiply },
+            { '/', Divide }
+        };
+
+        public static bool TryCalculate(string input, out int result)
+        {
+            result = 0;
+            string[] opsAndInts = input.Split(" ");
```
Comment (mmwoodfo) [inline Calculate/Calculate/Calculator.cs:21]: You should be using the Split method along side pattern matching. e.g.
`if (expression?.Split(" ") is [string operator1, [char mathmaticalOperator], string operand2])`

File: Calculate/Calculate.Tests/CalculateTests.cs [hunk 12]
```diff
@@ -0,0 +1,44 @@
+namespace Calculate.Tests
+{
+    [TestClass]
```
Comment (mmwoodfo) [inline Calculate/Calculate.Tests/CalculateTests.cs:3]: It would be nice to have some data-driven tests here to test more edge cases and numerical combinations
